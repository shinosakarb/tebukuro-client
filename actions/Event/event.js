// @flow
import { createAction } from 'redux-actions'
import Router from 'next/router'
import type { Dispatch } from 'redux'

import ActionsType from '../../constants/Actions'
import type { EventProps } from '../../types/Event'
// TODO: Enable ESLint and flow check after API implemented.
/* eslint-disable */
// flow-disable-nextline
import EventAPI from '../../api/Event'

const create = createAction(ActionsType.Event.createEvent)

export const createEvent = (params: EventProps) => (dispatch: Dispatch) => (
  EventAPI.create(params)
    .then((json) => {
      Router.replace(`/event/${json.id}`)
      return dispatch(create(json))
    })
    .catch(error => dispatch(create(error)))
)
