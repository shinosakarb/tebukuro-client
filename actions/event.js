// @flow
import { createAction } from 'redux-actions'
import Router from 'next/router'
import type { Dispatch } from 'redux'
import ActionsType from '../constants/Actions'
import { event, participant } from '../api'
import type { EventProps } from '../types/Event'

const create = createAction(ActionsType.Event.createEvent, event.create)

// TODO: Return reject in thunk.
export const createEvent = (params: EventProps) => (dispatch: Dispatch, getState: Function) => (
  dispatch(create(params)).then((res) => {
    const { id } = getState().event
    !res.error && Router.replace(`/event/show?id=${id}`, `/event/${id}`)
  })
)

export const fetchEvent = createAction(ActionsType.Event.fetchEvent, event.find)
export const registerForEvent = createAction(ActionsType.Event.registerForEvent, participant.create)
