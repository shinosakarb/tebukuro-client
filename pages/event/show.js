// @flow
import React from 'react'

export default () => (
  <div>
    <h3>This is the event page!</h3>
  </div>
)
