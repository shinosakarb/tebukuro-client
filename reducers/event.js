// @flow
import { handleActions } from 'redux-actions'
import { normalize } from 'normalizr'
import { Map, List } from 'immutable'

import Actions from '../constants/Actions'
import EventSchema from '../schemas/event'

export const eventInitialState = new Map({
  entityId: 0,
  entities: new Map(),
  errors: new List(),
})

const setEvent = {
  next: (state, action) => {
    const normalizedPayload = normalize(action.payload, EventSchema)

    return state.merge({
      entityId: normalizedPayload.result,
      entities: normalizedPayload.entities.event,
      errors: [],
    })
  },
  throw: (state, action) => state.merge({ errors: action.payload.errors }),
}

const mergeParticipant = {
  next: (state, action) => {
    const eventId = state.get('entityId').toString()
    const participantId = action.payload.id

    return state.updateIn(
      ['entities', eventId, 'participants'],
      participants => participants.push(participantId),
    )
  },
}

const eventReducerMap = {
  [Actions.Event.createEvent]: setEvent,
  [Actions.Event.fetchEvent]: setEvent,
  [Actions.Event.registerForEvent]: mergeParticipant,
}

export default handleActions(eventReducerMap, eventInitialState)
