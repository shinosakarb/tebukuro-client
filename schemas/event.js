import { schema } from 'normalizr'
import ParticipantSchema from './participant'

const EventSchema = new schema.Entity('event', {
  participants: [ParticipantSchema],
})

export default EventSchema
