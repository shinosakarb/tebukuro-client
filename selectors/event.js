// @flow
import { createSelector } from 'reselect'

const getEventEntities = state => state.event.get('entities')
const getEventErrors = state => state.event.get('errors')

export const getEventId = (state: Object) => state.event.get('entityId')
export const getEventErrorsArray = createSelector(getEventErrors, errors => errors.toArray())
export const getCurrentEvent = createSelector(
  [getEventId, getEventEntities],
  (id, entities) => {
    const entitiesObject = entities.get(id.toString()).toObject()
    return {
      ...entitiesObject,
      participants: entitiesObject.participants.toArray(),
    }
  },
)
export const getHasWaitlist =
  createSelector(getCurrentEvent, event => event.participants.length >= event.quota)
export const getHasNotFoundError =
  createSelector(getEventErrors, errors => errors.indexOf('Not Found') !== -1)
