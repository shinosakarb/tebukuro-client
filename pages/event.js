// @flow
import React from 'react'
import Event from '../containers/Event'
import type { EventId } from '../types/Event'

export default (props: {url: {query: EventId}}) => (
  <div>
    <h3>This is the event page!</h3>
    <Event eventId={props.url.query.id} />
  </div>
)
