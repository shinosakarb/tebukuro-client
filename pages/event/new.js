// @flow
import React from 'react'
import withRedux from 'next-redux-wrapper'
import createStore from '../../store'
import { createEvent } from '../../actions/event'
import { getEventErrorsArray } from '../../selectors/event'
import EventFormComponent from '../../components/EventForm'

type EventFormProps = {
  createEvent: Function,
  errors: ?string[]
}

export const NewEvent = (props: EventFormProps) => (
  <div>
    <h3>This is the event form page!</h3>
    { props.errors &&
      <ul>
        { props.errors.map(error =>
          <li>{ error }</li>)}
      </ul>
    }
    <EventFormComponent onSubmit={props.createEvent} />
  </div>
)

const mapDispatchToProps = { createEvent }
const mapStateToProps = state => (
  { errors: getEventErrorsArray(state) }
)

const connectProps = {
  createStore,
  mapStateToProps,
  mapDispatchToProps,
}

export default withRedux(connectProps)(NewEvent)
