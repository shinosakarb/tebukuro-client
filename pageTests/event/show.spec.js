import { shallow, mount } from 'enzyme'
import shallowToJson from 'enzyme-to-json'
import { ShowEvent } from '../../pages/event/show'
import EventParams from '../../factories/Event'
import ParticipantParams from '../../factories/Participant'

jest.mock('../../components/Event', () => 'EventComponent')
jest.mock('../../actions/event', () => ({
  fetchEvent: () => {},
}))

const testProps = {
  url: { query: { id: EventParams.event1.id } },
  event: EventParams.event1,
  hasNotFoundError: false,
  hasWaitlist: false,
  participants: Object.values(ParticipantParams),
  fetchEvent: jest.fn(),
  registerForEvent: jest.fn(),
}

describe('ShowEvent', () => {
  beforeEach(() => {
    testProps.fetchEvent.mockReset()
  })

  describe('when available to participate', () => {
    it('renders the event page with participate text.', () => {
      const page = shallow(<ShowEvent {...testProps} />)
      const tree = shallowToJson(page)

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when having a waiting list', () => {
    it('renders the event page wtih waitlisted text.', () => {
      const waitlistProps = { ...testProps, hasWaitlist: true }

      const page = shallow(<ShowEvent {...waitlistProps} />)
      const tree = shallowToJson(page)

      expect(tree).toMatchSnapshot()
    })
  })

  it('renders the page with error messages.', () => {
    const errorTestProps = { ...testProps, event: EventParams.errorEvent }
    const page = shallow(<ShowEvent {...errorTestProps} />)
    const tree = shallowToJson(page)

    expect(tree).toMatchSnapshot()
  })

  it('renders the page not found error page.', () => {
    const notFoundErrorProps = { ...testProps, hasNotFoundError: true }
    const page = shallow(<ShowEvent {...notFoundErrorProps} />)
    const tree = shallowToJson(page)

    expect(tree).toMatchSnapshot()
  })

  it('should call fetchEvent once with correct params', () => {
    mount(<ShowEvent {...testProps} />)

    expect(testProps.fetchEvent).toHaveBeenCalledTimes(1)
    expect(testProps.fetchEvent).toBeCalledWith({ id: EventParams.event1.id })
  })
})
