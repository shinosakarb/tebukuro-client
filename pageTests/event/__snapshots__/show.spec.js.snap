// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowEvent renders the event page. 1`] = `
<div>
  <h3>
    This is the event page!
  </h3>
  <EventComponent
    event={
      Object {
        "description": "This is the first event.",
        "id": 1,
        "name": "event1",
        "participants": Array [],
      }
    }
  />
  <ParticipantForm
    eventId={1}
    onSubmit={[Function]}
  />
</div>
`;

exports[`ShowEvent renders the page not found error page. 1`] = `
<Error
  statusCode="404"
/>
`;

exports[`ShowEvent renders the page with error messages. 1`] = `
<div>
  <ul>
    <li>
      nameを入力してください
    </li>
    <li>
      nameは１０文字以内です
    </li>
  </ul>
  <h3>
    This is the event page!
  </h3>
  <EventComponent
    event={
      Object {
        "errors": Array [
          "nameを入力してください",
          "nameは１０文字以内です",
        ],
      }
    }
  />
  <ParticipantForm
    onSubmit={[Function]}
  />
</div>
`;
